### PR TITLE
Cosim/fem dem

### PR DESCRIPTION
**Pull request by KlausBSautter**
added dem-fem tests and a dem-wrapper to the CoSimApp.

waiting for #5297 to be merged first

### CODE REVIEW

File: applications/CoSimulationApplication/tests/dem_fem_cable_net/cableNetFEMStructuralMaterials.json [hunk 1]
```diff
@@ -0,0 +1,20 @@
+{
+	"properties" : [{
+		"model_part_name" : "Parts_trusses",
```
Comment (philbucher) [inline applications/CoSimulationApplication/tests/dem_fem_cable_net/cableNetFEMStructuralMaterials.json]:
```suggestion
		"model_part_name" : "Structure.Parts_trusses",
```

File: applications/CoSimulationApplication/tests/dem_fem_cable_net/ProjectParametersFEM.json [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+		},
+		"line_search"                         : false,
+		"convergence_criterion"               : "residual_criterion",
+		"displacement_relative_tolerance"     : 1e-9,
+		"displacement_absolute_tolerance"     : 1e-9,
+		"residual_relative_tolerance"         : 1e-9,
+		"residual_absolute_tolerance"         : 1e-9,
+		"max_iteration"                       : 400,
+		"problem_domain_sub_model_part_list"  : ["Parts_trusses"],
+		"processes_sub_model_part_list"       : ["DISPLACEMENT_dirichXYZ","RESULT_middleNode"],
```
Comment (philbucher) [inline applications/CoSimulationApplication/tests/dem_fem_cable_net/ProjectParametersFEM.json]: can you please remove `processes_sub_model_part_list` and `problem_domain_sub_model_part_list`, those should no longer be needed

File: applications/CoSimulationApplication/tests/dem_fem_cable_net/ProjectParametersDEM.json [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+"MaxTimeStep"                      : 5e-5,
+"FinalTime"                        : 0.01,
+"ControlTime"                      : 4.0,
+"NeighbourSearchFrequency"         : 1,
+
+"GraphExportFreq"                  : 10e1,
+"VelTrapGraphExportFreq"           : 10e1,
+"OutputTimeStep"                   : 10e1,
+"PostBoundingBox"                  : false,
+"PostDisplacement"                 : true,
```
Comment (philbucher) [inline applications/CoSimulationApplication/tests/dem_fem_cable_net/ProjectParametersDEM.json]: needed for the test?

File: applications/CoSimulationApplication/python_scripts/coupling_operations/create_point_load_model_part.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    def FinalizeCouplingIteration(self):
+        pass
+
+    def Execute(self):
+        pass
+
+    def PrintInfo(self):
+        pass
+
+    def Check(self):
```
Comment (philbucher) [inline applications/CoSimulationApplication/python_scripts/coupling_operations/create_point_load_model_part.py]: please remove the empty calls

File: applications/CoSimulationApplication/python_scripts/solver_wrappers/kratos/dem_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+            dem_analysis_module = DEMAnalysisStageWithWorkingDir
+
+        return dem_analysis_module(self.model, self.project_parameters)
+
+    def SolveSolutionStep(self):
+        super(DEMWrapper,self).SolveSolutionStep()
+
+        # move the rigid wall object in the dem mp w.r.t. the current displacement and velocities
+        for mp_name in self.settings["settings"]["move_mesh_model_part"].GetStringArray():
```
Comment (philbucher) [inline applications/CoSimulationApplication/python_scripts/solver_wrappers/kratos/dem_wrapper.py]: better to extract it once and the use it, will be faster

**Review by philbucher — COMMENTED** [review]
do you need `cableNetDEM_Clusters.mdpa`?
It is empty

**Comment by KlausBSautter** [comment]
> do you need `cableNetDEM_Clusters.mdpa`?
> It is empty

no

**Comment by KlausBSautter** [comment]
@philbucher thanks for the comments. I fixed them and added a check for the current numbers of conditions

File: applications/CoSimulationApplication/python_scripts/solver_wrappers/kratos/dem_wrapper.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+            dem_analysis_module = DEMAnalysisStageWithWorkingDir
+
+        return dem_analysis_module(self.model, self.project_parameters)
+
+    def Initialize(self):
+        super(DEMWrapper,self).Initialize()
+
+        # save nodes in model parts which need to be moved while simulating
+        self.list_of_nodes_in_move_mesh_model_parts = []
```
Comment (philbucher) [inline applications/CoSimulationApplication/python_scripts/solver_wrappers/kratos/dem_wrapper.py]:
```suggestion
        self.list_of_nodes_in_move_mesh_model_parts = [self.model[mp_name].Nodes for mp_name in self.settings["settings"]["move_mesh_model_part"].GetStringArray()]
```

**Review by philbucher — APPROVED** [review]

**Pull request merged**